src/platform_impl/macos/view.rs

```diff
@@ -433,6 +501,34 @@ extern "C" fn set_marked_text(
             marked_text.initWithString(string);
         };
         *marked_text_ref = marked_text;
+
+        // Update ViewState with new marked text
+        let state_ptr: *mut c_void = *this.get_ivar("winitState");
+        let state = &mut *(state_ptr as *mut ViewState);
+        let preedit_string = id_to_string_lossy(string);
+
+        // Notify IME is active if application still doesn't know it.
+        if state.ime_state == ImeState::Disabled {
+            state.input_source = current_input_source();
+            AppState::queue_event(EventWrapper::StaticEvent(Event::WindowEvent {
+                window_id: WindowId(get_window_id(state.ns_window)),
+                event: WindowEvent::IME(IME::Enabled),
+            }));
+        }
+
+        let cursor_start = preedit_string.len();
+        let cursor_end = preedit_string.len();
+        state.ime_state = ImeState::Preedit;
+
+        // Send WindowEvent for updating marked text
+        AppState::queue_event(EventWrapper::StaticEvent(Event::WindowEvent {
+            window_id: WindowId(get_window_id(state.ns_window)),
+            event: WindowEvent::IME(IME::Preedit(
+                preedit_string,
+                Some(cursor_start),
+                Some(cursor_end),
```

Comment:
The purpose is that the position of your cursor is defined by your IME. For example on Wayland you can move your cursor around in preedit text, however it's done entirely by IME. So cursor position is to indicate where your cursor is currently is and where its end. Since if start != end you likely have a so called marked text(selection). The position of the cursor in preedit determines how it should look in the end.

I still have a question who is responsible to update it on macOS, since I can move cursor on preedit string in terminal.app, but not with winit, since when I press left/rigth nothing changes, however in terminal app it seems like moving.

Maybe the user should set cursor position itself on macOS, but at this information is essential for Wayland IME and is the only way to obtain current cursor position in preedit string, so you can draw cursor in the right place.

Does it clarify it for you?

Reply:
Hmm, I think downstream (e.g. `alacritty`/`winit`, not AppKit) is expected to render things differently, including the cursor/selected text. At first I thought `firstRectForCharacterRange:actualRange:` could be used to help tracking the currently focused characters, but then again, its value is cached, so probably not.

I actually think the `selectedRange` parameter in `setMarkedText:selectedRange:replacementRange:` could do the trick though.

Reply:
I mean, the current information this API exposes is already enough for alacritty to draw everything. At least on Wayland it looks like in any other app from my testing.